### ISSUE TITLE
Recomendación

### ISSUE THREAD

**Issue by EmaGP** [EmaGP/Mini-Proyectos#1]
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_percentage_error, precision_score, recall_score
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics.pairwise import cosine_similarity

# Carga de datos
data = pd.read_csv('retail_data.csv')

# Preprocesamiento de datos
data['product_features'] = data['product_name'] + ' ' + data['product_description']

# Codificar etiquetas de clientes y productos para modelos
le_customer = LabelEncoder()
data['customer_id'] = le_customer.fit_transform(data['customer_id'])

le_product = LabelEncoder()
data['product_id'] = le_product.fit_transform(data['product_id'])

# Crear un TfidfVectorizer para transformar las características del producto en vectores TF-IDF
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(data['product_features'])

# Calcular la similitud coseno entre productos
similarity_matrix = cosine_similarity(tfidf_matrix)

# Combinar TF-IDF con los datos de productos
tfidf_df = pd.DataFrame(tfidf_matrix.toarray(), columns=vectorizer.get_feature_names_out())
data = pd.concat([data, tfidf_df], axis=1)

# Separar características (X) de la cantidad de compras (y)
X = data.drop(columns=['customer_id', 'product_id', 'quantity', 'product_name', 'product_description', 'date', 'product_features'])
y = data['quantity']

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Entrenar un modelo de regresión lineal
model = LinearRegression()
model.fit(X_train, y_train)

# Realizar predicciones sobre la cantidad comprada
y_pred = model.predict(X_test)

# Evaluar el modelo
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
mape = mean_absolute_percentage_error(y_test, y_pred)

print(f'MAE (Mean Absolute Error): {mae:.4f}')
print(f'MSE (Mean Squared Error): {mse:.4f}')
print(f'R2 (Coeficiente de determinación): {r2:.4f}')
print(f'MAPE (Mean Absolute Percentage Error): {mape:.4f}')

# Función para recomendar productos
def recommend_products(product_id, k=5):
    # Obtener el índice del producto
    product_index = data[data['product_id'] == product_id].index[0]
    
    # Obtener similitudes para el producto
    similarity_scores = similarity_matrix[product_index]
    
    # Crear una lista de índices y sus puntajes de similitud
    indices_scores = list(enumerate(similarity_scores))
    
    # Ordenar productos por similitud (mayor a menor)
    indices_scores_sorted = sorted(indices_scores, key=lambda x: x[1], reverse=True)
    
    # Obtener los índices de los k productos más similares (excluyendo el producto dado)
    recommended_indices = [i for i, _ in indices_scores_sorted[1:k+1]]
    
    # Obtener los IDs de los productos recomendados
    recommended_products = data.iloc[recommended_indices]['product_id'].tolist()
    
    return recommended_products

# Calcular precisión y recall
def calculate_precision_recall(y_true, y_pred):
    # Calcular precisión y recall
    precision = precision_score(y_true, y_pred, average='binary')
    recall = recall_score(y_true, y_pred, average='binary')
    
    return precision, recall

# Ejemplo de uso: Recomendaciones para un producto dado
product_id = data['product_id'].iloc[0]  # Producto específico para el que quieres recomendaciones
k = 5  # Número de recomendaciones deseadas
recommended_products = recommend_products(product_id, k)
print(f'Productos recomendados para el producto ID {product_id}: {recommended_products}')

# Para calcular la precisión y recall, se necesita obtener los productos comprados por un cliente (en el conjunto de prueba)
y_true = list(X_test['product_id'])  # Los productos que el cliente compró

# Calcular precision@k
precision = precision_score(y_true, recommended_products, average='binary')
recall = recall_score(y_true, recommended_products, average='binary')

print(f'Precisión: {precision:.4f}')
print(f'Recall: {recall:.4f}')

# Calcular precision@k
def precision_at_k(y_true, y_pred, k):
    # Ordenar las predicciones
    y_pred_sorted = sorted(y_pred, key=lambda x: x[1], reverse=True)
    
    # Calcular precision@k
    true_positives = len([x for x in y_pred_sorted[:k] if x in y_true])
    precision = true_positives / k
    
    return precision

# Calcular precision@k para los productos recomendados
precision_at_k = precision_at_k(y_true, recommended_products, k)
print(f'Precision@k: {precision_at_k:.4f}')
